rust/arrow/src/ipc/reader.rs

```diff
@@ -521,7 +524,111 @@ impl<R: Read + Seek> Reader<R> {
     }
 }
 
-impl<R: Read + Seek> RecordBatchReader for Reader<R> {
+impl<R: Read + Seek> RecordBatchReader for FileReader<R> {
+    fn schema(&mut self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    fn next_batch(&mut self) -> Result<Option<RecordBatch>> {
+        self.next()
+    }
+}
+
+/// Arrow Stream reader
+pub struct StreamReader<R: Read> {
+    /// Buffered stream reader
+    reader: BufReader<R>,
+    /// The schema that is read from the stream's first message
+    schema: Arc<Schema>,
+    /// An indicator of whether the strewam is complete.
+    ///
+    /// This value is set to `true` the first time the reader's `next()` returns `None`.
+    finished: bool,
+}
+
+impl<R: Read> StreamReader<R> {
+    /// Try to create a new stream reader
+    ///
+    /// The first message in the stream is the schema, the reader will fail if it does not
+    /// encounter a schema.
+    /// To check if the reader is done, use `is_finished(self)`
+    pub fn try_new(reader: R) -> Result<Self> {
+        let mut reader = BufReader::new(reader);
+        // determine metadata length
+        let mut meta_size: [u8; 4] = [0; 4];
+        reader.read_exact(&mut meta_size)?;
+        let meta_len = u32::from_le_bytes(meta_size);
+
+        let mut meta_buffer = vec![0; meta_len as usize];
+        reader.read_exact(&mut meta_buffer)?;
+
+        let vecs = &meta_buffer.to_vec();
+        let message = ipc::get_root_as_message(vecs);
+        // message header is a Schema, so read it
+        let ipc_schema: ipc::Schema = message
+            .header_as_schema()
+            .expect("Arrow Stream must start with a schema");
```

Comment:
This method already returns a `Result` so shouldn't we propagate this up instead of using `expect`?

Reply:
Done